### PR TITLE
feat(cpf-matcher): Authenticator SPI de matching por CPF (auto-link + auto-heal)

### LINKED ISSUE #1: feat(keycloak): Authenticator SPI customizado para matching por CPF (auto-link + auto-heal)
## Contexto

A validação do Identity Brokering com gov.br ([uniplus-api PR #216](https://github.com/unifesspa-edu-br/uniplus-api/pull/216)) confirmou empiricamente o que a discussão arquitetural já apontava: para a plataforma Uni+, **CPF é o único identificador canônico estável** entre gov.br e LDAP institucional. Email não serve como chave alternativa porque parte significativa dos servidores Unifesspa cadastra a conta gov.br com email pessoal (Gmail/Hotmail), não institucional.

Adicionalmente, o LDAP institucional Unifesspa tem um bug histórico de dado: parte dos `brPersonCPF` está armazenado com 10 dígitos (zero à esquerda truncado, provavelmente por carga inicial via `int` em algum sistema legado). gov.br retorna sempre 11 dígitos no claim `sub`. Matching ingênuo por CPF falha para esses entries.

A solução é um **Authenticator Java SPI customizado no Keycloak** que substitua o flow padrão de "Detect Existing Broker User" (que casa por email) por matching por atributo `cpf` com tolerância para o bug do LDAP malformado.

O ambiente de teste local já existe (Story uniplus-api#217, mergeada na PR #216): OpenLDAP sintético com 3 entries reproduzindo o bug + 7 entries normais.

Esta é a **primeira Story deste repositório** — `uniplus-keycloak-providers` foi criado para hospedar SPI providers customizados do Keycloak Uni+ (Authenticators, mappers, event listeners). Estrutura multi-módulo Maven antecipa futuros providers como siblings.

## Objetivo

Implementar Authenticator SPI Java que, durante o flow `first broker login` do Keycloak, ao receber um broker context do gov.br:

1. **Tenta match com CPF de 11 dígitos** (formato canônico gov.br)
2. **Fallback: tenta match com CPF de 10 dígitos** (LDAP malformado, sem zero à esquerda)
3. **Se encontrar:** linka a identidade gov.br ao user existente + **auto-heal** (atualiza atributo `cpf` no Keycloak para 11 dígitos canônicos)
4. **Se não encontrar:** delega para o flow padrão (criar user novo)

### 1. Estrutura inicial do repositório

```
uniplus-keycloak-providers/
├── README.md                                              (visão geral do repo)
├── LICENSE                                                (Apache-2.0)
├── .gitignore                                             (Maven + IDE padrão)
├── CONTRIBUTING.md                                        (padrões do projeto)
├── pom.xml                                                (parent pom multi-module)
└── cpf-matcher/
    ├── pom.xml
    ├── src/main/java/br/edu/unifesspa/uniplus/keycloak/
    │   ├── CpfMatcherAuthenticator.java                   (lógica core)
    │   └── CpfMatcherAuthenticatorFactory.java            (registro do provider)
    ├── src/main/resources/META-INF/services/
    │   └── org.keycloak.authentication.AuthenticatorFactory
    │       (linha única apontando pra factory)
    ├── src/test/java/br/edu/unifesspa/uniplus/keycloak/
    │   └── CpfMatcherAuthenticatorTest.java               (JUnit + Mockito)
    └── README.md
```

### 2. Dependências Maven

- `org.keycloak:keycloak-server-spi` (provided)
- `org.keycloak:keycloak-server-spi-private` (provided)
- `org.keycloak:keycloak-services` (provided)
- `org.junit.jupiter:junit-jupiter` (test)
- `org.mockito:mockito-core` (test)

Versão alvo: **Keycloak 26.5** (mesma do compose do uniplus-api).
Versão Java: **21 LTS**.

### 3. Lógica do Authenticator

```java
public class CpfMatcherAuthenticator extends AbstractIdpAuthenticator {

    @Override
    protected void actionImpl(AuthenticationFlowContext ctx,
                              SerializedBrokeredIdentityContext serializedCtx,
                              BrokeredIdentityContext brokerCtx) {
        String cpfFromIdP = brokerCtx.getId();  // claim "sub" do gov.br

        // 1. Tenta matching com 11 dígitos (canônico)
        UserModel user = findUserByCpf(ctx, cpfFromIdP);

        // 2. Fallback: 10 dígitos (LDAP malformado)
        if (user == null && cpfFromIdP.length() == 11 && cpfFromIdP.startsWith("0")) {
            String stripped = cpfFromIdP.substring(1);
            user = findUserByCpf(ctx, stripped);
            if (user != null) {
                logger.info("CPF matching: encontrado via fallback 10-dig (LDAP malformado)");
                // Auto-heal: atualiza atributo para canônico
                user.setSingleAttribute("cpf", cpfFromIdP);
            }
        }

        if (user != null) {
            ctx.setUser(user);
            ctx.success();
        } else {
            ctx.attempted();  // delega ao próximo authenticator do flow
        }
    }

    private UserModel findUserByCpf(AuthenticationFlowContext ctx, String cpf) {
        return ctx.getSession().users()
            .searchForUserByUserAttributeStream(ctx.getRealm(), "cpf", cpf)
            .findFirst()
            .orElse(null);
    }
}
```

> Pseudocódigo — assinatura exata a confirmar contra Keycloak 26.5 SPI.

#### Build

```bash
mvn clean package
# Gera cpf-matcher/target/cpf-matcher-1.0.0.jar
```

#### Deploy no compose do uniplus-api

Volume mount no `docker-compose.yml` do **uniplus-api** apontando para o JAR compilado deste repo. Padrão sugerido: clonar este repo lado a lado com `uniplus-api`, e montar via path relativo. Detalhes na implementação (PR de uniplus-api estende compose após este SPI estar buildado).

#### Configuração do realm

Estender `setup-keycloak-dev.sh` no uniplus-api:

1. Clonar o flow built-in `first broker login` para `first broker login com cpf`
2. Inserir execution `cpf-matcher-authenticator` ANTES de `Confirm Link Existing Account`
3. Atualizar IdP `govbr` para usar o flow customizado (`firstBrokerLoginFlowAlias`)
4. Idempotente: detectar se já existe e re-criar se config mudou

### 5. Documentação

- `README.md` raiz do repo: propósito, lista de providers, como contribuir
- `cpf-matcher/README.md`: visão técnica, build, deploy, debug
- `CONTRIBUTING.md`: padrões de commit (conventional commits pt-BR), branch naming, sub-issues

### Build e deploy

- [ ] `mvn clean package` no root gera todos os JARs sem erros
- [ ] JAR `cpf-matcher-1.0.0.jar` é carregado pelo Keycloak (logs mostram registro do provider)
- [ ] Setup script no uniplus-api configura o flow customizado de forma idempotente
- [ ] Re-rodar setup 2x produz o mesmo estado (sem duplicação de execuções)

### Comportamento funcional (validação contra OpenLDAP sintético)

Os 3 entries malformados do LDAP local (uniplus-api Story #217) servem de cenário de teste:

- `kevin.peixoto` — CPF canônico `07094871422`, LDAP `7094871422`
- `emanuelly.fernandes` — CPF canônico `03776203781`, LDAP `3776203781`
- `fernando.melo` — CPF canônico `03425754904`, LDAP `3425754904`

- [ ] **Cenário 1 — match direto (11 dig):** simular login broker com CPF `76323164930` (lara.almeida, 11 dig direto no LDAP) → matching encontra user, link criado, atributo `cpf` permanece `76323164930`
- [ ] **Cenário 2 — fallback (10 dig):** simular login broker com CPF `07094871422` (kevin, canônico) → matching tenta 11 dig (não encontra), tenta 10 dig (encontra), link criado, **atributo `cpf` no Keycloak fica `07094871422`** (auto-heal)
- [ ] **Cenário 3 — user novo:** simular login broker com CPF `12345678909` (não existente no LDAP) → matching não encontra → delega ao próximo authenticator do flow → cria user novo (comportamento default)

### Testes unitários (JUnit + Mockito)

- [ ] Cobre Cenário 1 com mock retornando user em 11 dig
- [ ] Cobre Cenário 2 com mock retornando vazio em 11 dig + user em 10 dig
- [ ] Cobre Cenário 3 com mock retornando vazio em ambos
- [ ] Cobre edge cases: CPF nulo, CPF com tamanho inválido (≠ 10 e ≠ 11), CPF não começando com 0 (não tenta fallback)
- [ ] `mvn test` passa todos os testes

### Documentação

- [ ] README do repo explica propósito, lista de providers, como contribuir
- [ ] README do `cpf-matcher` explica build, deploy, debug
- [ ] CONTRIBUTING.md alinhado aos padrões do projeto Uni+
- [ ] Logs do Authenticator (info-level) registram qual tentativa de matching funcionou

## Não está no escopo

- Migração one-time do LDAP institucional (depende do DIRSI, fora do controle do projeto)
- ADR atualização (será feita após validação completa, possivelmente em ADR sucessora dedicada apenas a essa decisão)
- Deploy do JAR em HML/PROD (Story de operação separada)
- Suporte a outros formatos malformados além do "10 dígitos sem zero à esquerda" (escopo limitado ao bug conhecido)
- Outros providers (event listeners, mappers, etc.) — entrarão como Stories futuras neste mesmo repo

## Referências

- [uniplus-api PR #216](https://github.com/unifesspa-edu-br/uniplus-api/pull/216) — IdP gov.br + OpenLDAP sintético (este SPI é a peça final do quebra-cabeça)
- [uniplus-api Story #215](https://github.com/unifesspa-edu-br/uniplus-api/issues/215) — Configuração inicial do gov.br como IdP
- [uniplus-api Story #217](https://github.com/unifesspa-edu-br/uniplus-api/issues/217) — OpenLDAP sintético (ambiente de teste deste SPI)
- [uniplus-docs ADR-029](https://github.com/unifesspa-edu-br/uniplus-docs/pull/108) — Identity Brokering (será atualizada após esta Story)
- Documentação Keycloak SPI: https://www.keycloak.org/docs/latest/server_development/index.html#_auth_spi
- Exemplo de Authenticator customizado: https://github.com/keycloak/keycloak/tree/main/examples/providers/authenticator

## Bloqueia

- Fechamento da decisão arquitetural sobre matching de CPF (potencialmente nova ADR substituindo/complementando ADR-029)
- Habilitação do gov.br como IdP de produção para servidores Unifesspa

### PR DESCRIPTION
**Pull request by marmota-alpina**
## Summary

Implementa o módulo `cpf-matcher` — Authenticator SPI para Keycloak 26.5 que faz **matching tolerante por CPF** entre identidades federadas (gov.br) e users existentes no realm, incluindo o caso do LDAP institucional Unifesspa com `brPersonCPF` truncado em 10 dígitos. Primeira Story funcional do repo.

## Componentes

| Tipo | Arquivo | Papel |
|---|---|---|
| Value object (record) | `CanonicalCpf.java` | Encapsula domínio CPF: validação no construtor, fallback `truncated()`, mascaramento LGPD |
| Authenticator | `CpfMatcherAuthenticator.java` | Orquestrador enxuto, delega ao `CanonicalCpf`, segue padrão `idp-detect-existing-broker-user` |
| Factory | `CpfMatcherAuthenticatorFactory.java` | Registro SPI singleton stateless |
| SPI registration | `META-INF/services/org.keycloak.authentication.AuthenticatorFactory` | Linha única apontando para a Factory |

## Como funciona

```
sub do gov.br (sempre 11 dig)
        │
        ▼
CanonicalCpf.from(sub)
   │
   ├── inválido → context.attempted() (delega)
   │
   └── válido ↓
       searchByAttribute("cpf", canonical)
       │
       ├── encontrou → registerExistingUser + attempted()
       │
       └── não, e startsWithZero:
              searchByAttribute("cpf", truncated)
              │
              ├── encontrou → applyAutoHeal + register + attempted
              │
              └── não       → attempted (delega)
```

Em todos os caminhos chama `context.attempted()` — segue padrão do `idp-detect-existing-broker-user` builtin que coloca `EXISTING_USER_INFO` no auth note e deixa o próximo executor (`Auto Set Existing User`) fazer o link efetivo.

## Padrões aplicados (atende ao critério de qualidade Java do Uni+)

- **SRP:** lógica de domínio (validação, parsing, normalização CPF) isolada em `CanonicalCpf`; Authenticator orquestra apenas.
- **Imutabilidade:** `CanonicalCpf` é `record` com validação no construtor.
- **Fail-fast:** construtor rejeita input inválido; `from()` retorna `Optional` para uso seguro sem try/catch.
- **LGPD:** `toString` mascarado por padrão, `masked()` para logs explícitos, **CPF nunca em texto plano nos logs**.
- **Performance:** lazy logging via parametrized strings (JBoss Logging), busca fallback só executada se canônico inicia com 0, sem alocação extra em hot path.
- **Testabilidade:** helpers privados pequenos, value object testável isoladamente, mocks só na fronteira (Keycloak SPI), AAA + naming expressivo `metodo_quandoCondicao_deveResultado`.

### `CanonicalCpfTest` (16 testes em 4 nested groups)
- **Constructor:** aceita 11 dig, aceita zero à esquerda, rejeita tamanhos inválidos (parametrizado), rejeita null, rejeita não-dígitos (parametrizado)
- **From:** input null/inválido/tamanho/não-dígito retorna `Optional.empty`; input válido retorna preenchido
- **Fallback:** `startsWithZero` + `truncated` com e sem zero, tamanho do truncated
- **Masking:** formato `***.***.***-NN`, `toString` sempre mascarado
- **Value semantics:** equals/hashCode por valor

### `CpfMatcherAuthenticatorTest` (19 testes em 5 nested groups)
- **FunctionalScenarios (3):** match direto 11 dig, fallback 10 dig com auto-heal, sem match
- **FallbackDecision (2):** não tenta fallback sem zero, não tenta fallback quando match direto sucedeu
- **InputValidation (4):** cpf nulo, em branco, tamanho errado, caracteres inválidos
- **AutoHeal (1):** captura argumentos e verifica que escreve canônico de 11 dig começando com '0'
- **FactoryContract (2):** `requiresUser=false`, `configuredFor=true`

## Build

```bash
mvn clean verify
```

- 35/35 testes verdes
- JAR: `cpf-matcher/target/cpf-matcher-1.0.0-SNAPSHOT.jar` (~7.4KB)
- `META-INF/services/org.keycloak.authentication.AuthenticatorFactory` registra a Factory

## Próximas Stories (fora desta PR)

1. **uniplus-api**: ajustar `docker-compose.yml` para montar o JAR do cpf-matcher em `/opt/keycloak/providers/` e estender `setup-keycloak-dev.sh` para configurar o flow customizado
2. **uniplus-api**: validação E2E contra OpenLDAP sintético (já existe na branch `feature/215-govbr-idp`)
3. **uniplus-docs**: revisar/dividir ADR-029 conforme decisão consolidada (uma decisão por ADR)

## Test plan

- [ ] `mvn clean verify` — 35/35 testes verdes
- [ ] Reviewer revisa `CanonicalCpf` (contratos de domínio: validação, imutabilidade, mascaramento)
- [ ] Reviewer revisa `CpfMatcherAuthenticator` (uso correto de AbstractIdpAuthenticator + EXISTING_USER_INFO)
- [ ] Reviewer confere logs e garante que CPF nunca aparece em texto plano
- [ ] Reviewer valida estrutura multi-module Maven

Closes #1

### CODE REVIEW

File: cpf-matcher/src/main/java/br/edu/unifesspa/uniplus/keycloak/cpfmatcher/CpfMatcherAuthenticator.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+            .filter(java.util.Objects::nonNull)
+            .map(user -> new MatchResult(user, /* viaFallback */ true));
+    }
+
+    private UserModel findUserByAttribute(AuthenticationFlowContext context,
+                                          String attributeName, String attributeValue) {
+        return context.getSession().users()
+            .searchForUserByUserAttributeStream(context.getRealm(), attributeName, attributeValue)
+            .findFirst()
+            .orElse(null);
```
[the remark this answers is not in the document]
Comment (marmota-alpina) [reply cpf-matcher/src/main/java/br/edu/unifesspa/uniplus/keycloak/cpfmatcher/CpfMatcherAuthenticator.java]: Endereçado em 8c756ed (rejeita matching ambíguo).

A busca agora usa `searchForUserByUserAttributeStream(...).limit(2).toList()` em ambos os caminhos (canônico e fallback). Quando `size() > 1`, o método retorna `Optional.empty()` e loga warn com CPF mascarado (`***.***.***-NN`) sem registrar `EXISTING_USER_INFO` — o flow segue como se não houvesse match, evitando associação incorreta entre identidade gov.br e conta. Limit 2 é suficiente para distinguir 0/1/"muitos" sem materializar streams grandes em caso de atributo mal indexado.

Mensagem de log: *"Recusando link automático para evitar associação incorreta"*.

Fica como follow-up (fora desta PR) decidir o tratamento operacional para colisões reais — provavelmente um relatório no setup-keycloak-dev para sinalizar duplicatas no LDAP institucional antes do go-live.

File: cpf-matcher/src/main/java/br/edu/unifesspa/uniplus/keycloak/cpfmatcher/CpfMatcherAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    /**
+     * Atualiza o atributo {@code cpf} do user para o formato canônico (11 dígitos).
+     * Em users LDAP-federados read-only, esta escrita afeta apenas o cache local
+     * do Keycloak — o LDAP institucional permanece com o valor truncado original
+     * (correção definitiva exige migração no próprio LDAP).
+     */
+    void applyAutoHeal(UserModel user, CanonicalCpf canonicalCpf) {
+        user.setSingleAttribute(CPF_USER_ATTRIBUTE, canonicalCpf.value());
```
[the remark this answers is not in the document]
Comment (marmota-alpina) [reply cpf-matcher/src/main/java/br/edu/unifesspa/uniplus/keycloak/cpfmatcher/CpfMatcherAuthenticator.java]: Endereçado em 8c756ed (auto-heal best-effort).

`applyAutoHeal` agora envolve `user.setSingleAttribute` em `try/catch (RuntimeException)`. Se o storage rejeitar a escrita (LDAP em `READ_ONLY`, indisponibilidade temporária, qualquer erro de provider) o método loga warn com CPF mascarado e o tipo da exceção, e prossegue — `registerExistingUser` é chamado normalmente em seguida e o login conclui.

Decisão importante registrada no Javadoc: em users LDAP-federados a escrita só afeta o cache local do Keycloak, não o LDAP de origem. A correção definitiva do bug do `brPersonCPF` truncado exige migração no LDAP institucional (memorando ao DIRSI já está rascunhado, fora desta PR).

**Review by jf2s — APPROVED** [review]
Aprovado.

Revisei a Story após o rebase em `main` (commits `17b42bd` + `228b254`):

**Domínio (`CanonicalCpf`)**
- Record imutável, validação no construtor, fail-fast com `from()` retornando `Optional`. ✅
- Mascaramento `***.***.***-NN` consistente em `toString` e `masked()` — CPF nunca em texto plano nos logs. ✅

**Authenticator**
- Orquestrador enxuto, segue padrão `idp-detect-existing-broker-user` (sempre `attempted()`, registra `EXISTING_USER_INFO` para o próximo executor). ✅
- Findings P1 do Codex endereçados no commit `228b254`:
  - **Matching ambíguo:** `searchUsersByAttribute` agora limita a 2 e rejeita `size() > 1` em ambos os caminhos (canônico e fallback truncado), com warn mascarado por LGPD.
  - **Auto-heal read-only:** `applyAutoHeal` envolto em `try/catch (RuntimeException)` — falhas de storage (LDAP `READ_ONLY`, indisponibilidade) não derrubam matching já bem-sucedido. Javadoc explica que escrita em users LDAP-federados afeta apenas cache local.
- Busca `.limit(2)` é eficiência defensiva contra atributo mal indexado — bem pensado.

**Testes (35)**
- Cobertura sólida: 16 em `CanonicalCpf` (constructor, from, fallback, masking, value semantics) + 19 em `CpfMatcherAuthenticator` (functional scenarios, fallback decision, input validation, auto-heal arg capture, factory contract).
- Naming `metodo_quandoCondicao_deveResultado` consistente com padrão Java do Uni+.
- AAA, mocks só na fronteira Keycloak SPI.

**Estrutura Maven**
- Multi-module pronto para receber outros providers no mesmo repo.
- `META-INF/services/org.keycloak.authentication.AuthenticatorFactory` registrado corretamente.

Observação para o follow-up (Story #218): tratamento operacional de colisões reais de CPF no LDAP institucional pode virar relatório no setup-keycloak-dev antes do go-live.
